api/v1/storage_autoscaler.go

```diff
@@ -0,0 +1,128 @@
+	// +kubebuilder:validation:Required
+	StorageCluster corev1.LocalObjectReference `json:"storageCluster,omitempty"`
+	// DeviceClass is the name of the device class for which the storage scaling is to be done.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="ssd"
+	DeviceClass string `json:"deviceClass,omitempty"`
+	// StorageCapacityLimit is the total aggregate capacity limit for the storage scaling for the specific deviceClass and storagecluster.
+	// +kubebuilder:validation:Required
+	StorageCapacityLimit resource.Quantity `json:"storageCapacityLimit,omitempty"`
+	// MaxOsdSize is the maximum size that Osd disk can be expanded to.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default="8Ti"
+	MaxOsdSize resource.Quantity `json:"maxOsdSize,omitempty"`
+	// StorageScalingThresholdPercent is the threshold percentage of the storage capacity that triggers the auto-scaling of the OSDs.
+	// Should be less than the OsdNearFullThresholdPercentage.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=70
+	StorageScalingThresholdPercent int `json:"storageScalingThresholdPercent,omitempty"`
+	// TimeoutSeconds is the time in seconds after which the storage auto-scaler will alert the user that the scaling operation has been failed.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=1800
+	TimeoutSeconds int `json:"timeoutSeconds,omitempty"`
+}
+
+type StorageAutoScalerPhase string
+
+const (
+	StorageAutoScalerPhaseNotStarted StorageAutoScalerPhase = "NotStarted"
+	StorageAutoScalerPhaseInProgress StorageAutoScalerPhase = "InProgress"
+	StorageAutoScalerPhaseSucceeded  StorageAutoScalerPhase = "Succeeded"
+	StorageAutoScalerPhaseFailed     StorageAutoScalerPhase = "Failed"
+)
+
+// TimestampedError is the error message with the time stamp.
+type TimestampedError struct {
+	// Message is the error message in case the storage scaling operation has failed.
+	Message string `json:"message,omitempty"`
+	// Timestamp is the time stamp when the error occurred.
+	Timestamp metav1.Time `json:"timestamp,omitempty"`
+}
+
+// StorageAutoScalerStatus defines the observed state of StorageAutoScaler
+type StorageAutoScalerStatus struct {
+	// Phase describes the Phase of StorageAutoScaler
+	// +kubebuilder:validation:Enum="";'NotStarted;InProgress;Failed;Succeeded
+	Phase StorageAutoScalerPhase `json:"phase,omitempty"`
+	// Error is the error message in case the storage scaling operation has failed.
+	// +optional
+	// +nullable
+	Error *TimestampedError `json:"error,omitempty"`
+	// LastExpansionStartTime is the time stamp of the last run start of the storage scaling
+	// +optional
+	LastExpansionStartTime metav1.Time `json:"lastExpansionStartTime,omitempty"`
+	// LastExpansionCompletionTime is the time stamp of the last run completion of the storage scaling
+	// +optional
+	LastExpansionCompletionTime metav1.Time `json:"lastExpansionCompletionTime,omitempty"`
```

Comment:
I'd suggest asking @nb-ohad and @iamniting whether they think these time fields should be pointers/nullable so that the output doesn't include the epoch date-time when time is equal to 0.

```suggestion
	// LastExpansionStartTime is the time stamp of the last run start of the storage scaling
	// +optional
	// +nullable
	LastExpansionStartTime *metav1.Time `json:"lastExpansionStartTime,omitempty"`
	// LastExpansionCompletionTime is the time stamp of the last run completion of the storage scaling
	// +optional
	// +nullable
	LastExpansionCompletionTime *metav1.Time `json:"lastExpansionCompletionTime,omitempty"`
```

Reply:
It depends on the use case. If someone else is going to read the time, I would like to have it as a pointer so the other operator or code understands whether it is set or not yet.

Reply:
I dont think so any one else is reading it, its for the ui

Reply:
okay then having it  nullable  make sense, updated it to pointer